File: src/lucy_orchestrator/channels/feishu_webhook.py
from __future__ import annotations

import json
import threading
from dataclasses import dataclass
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from typing import TYPE_CHECKING, Any

from ..exceptions import OrchestratorError
from .feishu import FeishuMessenger, parse_requirement_event

if TYPE_CHECKING:
    from ..orchestrator import Orchestrator


@dataclass
class FeishuWebhookSettings:
    repo_name: str
    base_branch: str = "main"
    worktree_path: str | None = None
    auto_clarify: bool = True
    auto_run_on_approve: bool = False
    auto_provision_worktree: bool = False
    repo_path: str | None = None
    worktrees_root: str | None = None
    branch_prefix: str = "agent"
    send_reply: bool = True
    allow_from: list[str] | None = None
    verification_token: str | None = None


class ProcessedMessageStore:
    def __init__(
        self, file_path: str | Path = ".orchestrator/feishu_seen_messages.json"
    ) -> None:
        self.file_path = Path(file_path)
        self._lock = threading.Lock()
        self._seen: set[str] = set()
        self._load()

    def contains(self, message_id: str) -> bool:
        with self._lock:
            return message_id in self._seen

    def add(self, message_id: str) -> None:
        with self._lock:
            if message_id in self._seen:
                return
            self._seen.add(message_id)
            self._persist()

    def _load(self) -> None:
        if not self.file_path.exists():
            return
        try:
            payload = json.loads(self.file_path.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            return
        if not isinstance(payload, list):
            return
        self._seen = {str(item) for item in payload}

    def _persist(self) -> None:
        self.file_path.parent.mkdir(parents=True, exist_ok=True)
        data = sorted(self._seen)
        self.file_path.write_text(
            json.dumps(data, indent=2, ensure_ascii=False), encoding="utf-8"
        )


class FeishuWebhookProcessor:
    def __init__(
        self,
        orchestrator: "Orchestrator",
        settings: FeishuWebhookSettings,
        messenger: FeishuMessenger | None = None,
        processed_store: ProcessedMessageStore | None = None,
    ) -> None:
        self.orchestrator = orchestrator
        self.settings = settings
        self.messenger = messenger
        self.processed_store = processed_store or ProcessedMessageStore()

    def process_payload(self, payload: dict[str, Any]) -> tuple[int, dict[str, Any]]:
        event_type = self._event_type(payload)
        if payload.get("type") == "url_verification":
            challenge = str(payload.get("challenge", ""))
            if not challenge:
                return HTTPStatus.BAD_REQUEST, {"error": "missing challenge"}
            return HTTPStatus.OK, {"challenge": challenge}

        if event_type and event_type != "im.message.receive_v1":
            return HTTPStatus.OK, {
                "status": "ignored",
                "reason": f"unsupported_event_type:{event_type}",
            }

        try:
            requirement = parse_requirement_event(payload)
        except OrchestratorError as exc:
            return HTTPStatus.BAD_REQUEST, {"error": str(exc)}

        if (
            self.settings.allow_from
            and requirement.user_id not in self.settings.allow_from
        ):
            return HTTPStatus.OK, {
                "status": "ignored",
                "reason": "sender_not_allowed",
                "user_id": requirement.user_id,
            }

        if self.processed_store.contains(requirement.message_id):
            return HTTPStatus.OK, {
                "status": "duplicate",
                "message_id": requirement.message_id,
            }

        try:
            task, reply_text = self.orchestrator.process_feishu_message(
                requirement=requirement,
                repo_name=self.settings.repo_name,
                base_branch=self.settings.base_branch,
                worktree_path=self.settings.worktree_path,
                auto_clarify=self.settings.auto_clarify,
                auto_run_on_approve=self.settings.auto_run_on_approve,
                auto_provision_worktree=self.settings.auto_provision_worktree,
                repo_path=self.settings.repo_path,
                worktrees_root=self.settings.worktrees_root,
                branch_prefix=self.settings.branch_prefix,
            )
        except Exception as exc:
            return HTTPStatus.INTERNAL_SERVER_ERROR, {
                "status": "error",
                "reason": "orchestrator_failed",
                "error": str(exc),
            }

        reply_sent = False
        if self.settings.send_reply and self.messenger is not None:
            try:
                self.messenger.send_text(requirement.chat_id, reply_text)
                reply_sent = True
            except Exception as exc:
                return HTTPStatus.INTERNAL_SERVER_ERROR, {
                    "status": "error",
                    "reason": "reply_send_failed",
                    "error": str(exc),
                    "task_id": task.task_id,
                }

        self.processed_store.add(requirement.message_id)
        return HTTPStatus.OK, {
            "status": "ok",
            "task_id": task.task_id,
            "task_state": task.state.value,
            "reply_sent": reply_sent,
        }

    def validate_token(self, payload: dict[str, Any]) -> bool:
        expected = self.settings.verification_token
        if not expected:
            return True

        token = payload.get("token")
        if not token:
            header = payload.get("header", {})
            if isinstance(header, dict):
                token = header.get("token")

        return str(token or "") == expected

    @staticmethod
    def _event_type(payload: dict[str, Any]) -> str:
        header = payload.get("header", {})
        if not isinstance(header, dict):
            return ""
        return str(header.get("event_type", ""))


def serve_feishu_webhook(
    processor: FeishuWebhookProcessor,
    *,
    host: str = "0.0.0.0",
    port: int = 18791,
) -> None:
    class _Handler(BaseHTTPRequestHandler):
        def do_GET(self) -> None:  # noqa: N802
            if self.path == "/health":
                self._json_response(HTTPStatus.OK, {"status": "ok"})
                return
            self._json_response(HTTPStatus.NOT_FOUND, {"error": "not_found"})

        def do_POST(self) -> None:  # noqa: N802
            content_length = self.headers.get("Content-Length", "0")
            try:
                body_bytes = self.rfile.read(int(content_length))
            except ValueError:
                self._json_response(
                    HTTPStatus.BAD_REQUEST, {"error": "invalid content length"}
                )
                return

            try:
                payload = json.loads(body_bytes.decode("utf-8"))
            except json.JSONDecodeError:
                self._json_response(HTTPStatus.BAD_REQUEST, {"error": "invalid json"})
                return

            if not isinstance(payload, dict):
                self._json_response(
                    HTTPStatus.BAD_REQUEST, {"error": "payload must be object"}
                )
                return

            if not processor.validate_token(payload):
                self._json_response(
                    HTTPStatus.FORBIDDEN, {"error": "invalid verification token"}
                )
                return

            code, response_payload = processor.process_payload(payload)
            self._json_response(code, response_payload)

        def log_message(self, format: str, *args: Any) -> None:  # noqa: A003
            return

        def _json_response(self, status: int, payload: dict[str, Any]) -> None:
            body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
            self.send_response(status)
            self.send_header("Content-Type", "application/json; charset=utf-8")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)

    server = ThreadingHTTPServer((host, port), _Handler)
    try:
        server.serve_forever()
    finally:
        server.server_close()

File: src/channels/feishu-webhook.ts
import { createServer, type IncomingMessage, type ServerResponse } from "node:http"
import { mkdir, readFile, rename, writeFile } from "node:fs/promises"
import { dirname, resolve } from "node:path"

import { errorCodeOf, OrchestratorError } from "../errors.js"
import { logError, logWarn } from "../logger.js"
import { Orchestrator } from "../orchestrator.js"
import { readObject } from "./feishu-core.js"
import { FeishuMessenger, parseRequirementEvent } from "./feishu.js"

export interface FeishuWebhookSettings {
  repoName: string
  baseBranch?: string
  worktreePath?: string
  autoClarify?: boolean
  autoRunOnApprove?: boolean
  autoProvisionWorktree?: boolean
  repoPath?: string
  worktreesRoot?: string
  branchPrefix?: string
  sendReply?: boolean
  allowFrom?: string[]
  verificationToken?: string
}

export class ProcessedMessageStore {
  private readonly filePath: string
  private seen = new Set<string>()
  private loaded = false
  private writeChain: Promise<void> = Promise.resolve()

  constructor(filePath = ".orchestrator/feishu_seen_messages.json") {
    this.filePath = resolve(filePath)
  }

  async has(messageId: string): Promise<boolean> {
    await this.load()
    return this.seen.has(messageId)
  }

  async add(messageId: string): Promise<void> {
    await this.withWriteLock(async () => {
      await this.load()
      if (this.seen.has(messageId)) {
        return
      }
      this.seen.add(messageId)
      await this.persist()
    })
  }

  private async load(): Promise<void> {
    if (this.loaded) {
      return
    }
    this.loaded = true

    try {
      const raw = await readFile(this.filePath, "utf-8")
      try {
        const payload = JSON.parse(raw)
        if (Array.isArray(payload)) {
          this.seen = new Set(payload.map(String))
          return
        }
        logWarn("Processed message store payload is not an array, starting with empty cache", {
          phase: "webhook.processed-store.load",
          filePath: this.filePath,
        })
      } catch (error) {
        logWarn("Failed to parse processed message store JSON, starting with empty cache", {
          phase: "webhook.processed-store.parse",
          filePath: this.filePath,
          error: error instanceof Error ? error.message : String(error),
        })
      }
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === "ENOENT") {
        return
      }
      logWarn("Failed to read processed message store, starting with empty cache", {
        phase: "webhook.processed-store.read",
        filePath: this.filePath,
        error: error instanceof Error ? error.message : String(error),
      })
    }
  }

  private async persist(): Promise<void> {
    await mkdir(dirname(this.filePath), { recursive: true })
    const tempPath = `${this.filePath}.tmp-${process.pid}-${Date.now()}`
    await writeFile(tempPath, `${JSON.stringify([...this.seen].sort(), null, 2)}\n`, "utf-8")
    await rename(tempPath, this.filePath)
  }

  private async withWriteLock<T>(operation: () => Promise<T>): Promise<T> {
    const run = this.writeChain.then(operation, operation)
    this.writeChain = run.then(() => undefined, () => undefined)
    return run
  }
}

export class FeishuWebhookProcessor {
  constructor(
    private readonly orchestrator: Orchestrator,
    private readonly settings: FeishuWebhookSettings,
    private readonly messenger?: FeishuMessenger,
    private readonly processedStore = new ProcessedMessageStore(),
  ) {}

  async validateToken(payload: Record<string, unknown>): Promise<boolean> {
    if (!this.settings.verificationToken) {
      return true
    }
    const token =
      typeof payload.token === "string"
        ? payload.token
        : typeof readObject(payload.header).token === "string"
          ? (readObject(payload.header).token as string)
          : ""
    return token === this.settings.verificationToken
  }

  async processPayload(payload: Record<string, unknown>): Promise<{ statusCode: number; payload: Record<string, unknown> }> {
    const type = typeof payload.type === "string" ? payload.type : ""
    if (type === "url_verification") {
      const challenge = String(payload.challenge ?? "")
      if (!challenge) {
        return { statusCode: 400, payload: { error: "missing challenge" } }
      }
      return { statusCode: 200, payload: { challenge } }
    }

    const eventType = String(readObject(payload.header).event_type ?? "")
    if (eventType && eventType !== "im.message.receive_v1") {
      return {
        statusCode: 200,
        payload: { status: "ignored", reason: `unsupported_event_type:${eventType}` },
      }
    }

    let requirement
    try {
      requirement = parseRequirementEvent(payload)
    } catch (error) {
      logWarn("Failed to parse Feishu webhook requirement payload", {
        phase: "webhook.parse-requirement",
        error: error instanceof Error ? error.message : String(error),
      })
      return {
        statusCode: 400,
        payload: { error: error instanceof Error ? error.message : String(error) },
      }
    }

    if (this.settings.allowFrom && this.settings.allowFrom.length > 0) {
      if (!this.settings.allowFrom.includes(requirement.userId)) {
        return {
          statusCode: 200,
          payload: {
            status: "ignored",
            reason: "sender_not_allowed",
            userId: requirement.userId,
          },
        }
      }
    }

    if (await this.processedStore.has(requirement.messageId)) {
      return {
        statusCode: 200,
        payload: { status: "duplicate", messageId: requirement.messageId },
      }
    }

    try {
      const { task, replyText } = await this.orchestrator.processFeishuMessage({
        requirement,
        repoName: this.settings.repoName,
        baseBranch: this.settings.baseBranch,
        worktreePath: this.settings.worktreePath,
        autoClarify: this.settings.autoClarify,
        autoRunOnApprove: this.settings.autoRunOnApprove,
        autoProvisionWorktree: this.settings.autoProvisionWorktree,
        repoPath: this.settings.repoPath,
        worktreesRoot: this.settings.worktreesRoot,
        branchPrefix: this.settings.branchPrefix,
      })

      let replySent = false
      if (this.settings.sendReply !== false && this.messenger) {
        await this.messenger.sendText(requirement.chatId, replyText)
        replySent = true
      }

      await this.processedStore.add(requirement.messageId)
      if (!task) {
        return {
          statusCode: 200,
          payload: {
            status: "draft",
            replySent,
          },
        }
      }

      return {
        statusCode: 200,
        payload: {
          status: "ok",
          taskId: task.taskId,
          taskState: task.state,
          replySent,
        },
      }
    } catch (error) {
      logError("Feishu webhook payload processing failed", error, { phase: "webhook.process-payload" })
      return {
        statusCode: 500,
        payload: {
          status: "error",
          error: error instanceof Error ? error.message : String(error),
          errorCode: errorCodeOf(error),
        },
      }
    }
  }
}

export function serveFeishuWebhook(
  processor: FeishuWebhookProcessor,
  options: {
    host?: string
    port?: number
  } = {},
): Promise<void> {
  const host = options.host ?? "0.0.0.0"
  const port = options.port ?? 18791

  const server = createServer(async (request, response) => {
    try {
      await handleRequest(processor, request, response)
    } catch (error) {
      logError("Unhandled Feishu webhook request failure", error, { phase: "webhook.serve-request" })
      json(response, 500, {
        error: error instanceof Error ? error.message : String(error),
        errorCode: errorCodeOf(error),
      })
    }
  })

  return new Promise((resolvePromise, rejectPromise) => {
    server.on("error", rejectPromise)
    server.listen(port, host, () => resolvePromise())
  })
}

async function handleRequest(
  processor: FeishuWebhookProcessor,
  request: IncomingMessage,
  response: ServerResponse,
): Promise<void> {
  if (request.method === "GET" && request.url === "/health") {
    json(response, 200, { status: "ok" })
    return
  }

  if (request.method !== "POST") {
    json(response, 404, { error: "not_found" })
    return
  }

  const body = await readBody(request)
  let payload: Record<string, unknown>
  try {
    const parsed = JSON.parse(body)
    if (!parsed || typeof parsed !== "object") {
      throw new OrchestratorError("payload must be object")
    }
    payload = parsed as Record<string, unknown>
  } catch (error) {
    logWarn("Invalid JSON in Feishu webhook request body", {
      phase: "webhook.parse-body",
      error: error instanceof Error ? error.message : String(error),
    })
    json(response, 400, { error: "invalid json" })
    return
  }

  const valid = await processor.validateToken(payload)
  if (!valid) {
    json(response, 403, { error: "invalid verification token" })
    return
  }

  const result = await processor.processPayload(payload)
  json(response, result.statusCode, result.payload)
}

function readBody(request: IncomingMessage): Promise<string> {
  return new Promise((resolvePromise, rejectPromise) => {
    const chunks: Buffer[] = []
    request.on("data", (chunk) => chunks.push(Buffer.from(chunk)))
    request.on("end", () => resolvePromise(Buffer.concat(chunks).toString("utf-8")))
    request.on("error", rejectPromise)
  })
}

function json(response: ServerResponse, statusCode: number, payload: Record<string, unknown>): void {
  const body = JSON.stringify(payload)
  response.statusCode = statusCode
  response.setHeader("content-type", "application/json; charset=utf-8")
  response.end(body)
}
